src/command/mod.rs

```diff
@@ -1,32 +1,90 @@
+use std::time::Duration;
+
+use tokio::sync::oneshot;
+
 use crate::codec::Token;
+use crate::server::Context;
+use crate::storage::StorageCommand;
 
 mod command;
 pub use command::{Command, CommandError};
 
 /// CommandProcessor is responsible for taking a group of tokens, executing them,
 /// and returning the result.
-pub struct CommandProcessor {}
+pub struct CommandProcessor {
+    context: Context,
+}
 
 impl CommandProcessor {
-    pub fn execute_command(&self, command: &Command) -> Vec<Token> {
+    pub fn new(context: Context) -> Self {
+        Self { context }
+    }
+
+    pub async fn execute_command(&self, command: &Command) -> Vec<Token> {
         match command {
             Command::Echo(t) => vec![Token::BulkString(Some(t.clone()))],
             Command::Command => vec![Token::BulkString(Some("ECHO".bytes().collect()))],
+            Command::Get(key) => {
+                let cmd = StorageCommand::Get(key.clone());
+                let (tx, rx) = oneshot::channel();
+                let res = self
+                    .context
+                    .storage_queue
+                    .send_timeout((cmd, tx), Duration::from_millis(1_000))
+                    .await;
+
+                if let Err(_) = res {
+                    return error_response("timeout while sending to storage".to_string());
+                }
+
+                match rx.await {
+                    Ok(Ok(None)) => vec![Token::BulkString(None)],
+                    Ok(Ok(Some(value))) => vec![Token::BulkString(Some(value))],
+                    Ok(Err(_)) => error_response("internal storage error".to_string()),
+                    Err(_) => error_response("no response from storage".to_string()),
+                }
+            }
+            Command::Set(key, value) => {
+                let cmd = StorageCommand::Set(key.clone(), value.clone());
+                let (tx, rx) = oneshot::channel();
+                let res = self
+                    .context
+                    .storage_queue
+                    .send_timeout((cmd, tx), Duration::from_millis(1_000))
+                    .await;
+
+                if let Err(_) = res {
+                    return error_response("timeout while sending to storage".to_string());
+                }
+
+                match rx.await {
+                    Ok(Ok(None)) => vec![Token::SimpleString("OK".to_string())],
+                    Ok(Ok(Some(value))) => vec![Token::BulkString(Some(value))],
+                    Ok(Err(_)) => error_response("internal storage error".to_string()),
+                    Err(_) => error_response("no response from storage".to_string()),
+                }
+            }
         }
     }
 }
 
+fn error_response(msg: String) -> Vec<Token> {
```

Comment:
Consider changing this to `&str`, since it looks like you only ever call this function with static string slices.

Reply:
I am using `String` here since `Token::Error` accepts a `String`, so I figured I'd pass in the `String` directly instead of doing a copy inside (so if you already have a `String` you don't end up with possibly an extra copy). This will be something I revisit in the future when I try to make more things use references, but for now I think I will leave this

Reply:
Quite right. Today Mikkel realizes that this was a bad suggestion.

Reply:
It was a good suggestion in also that it made me and my pairing partners discuss what the structure is and whether or not a copy makes sense and where we'd like to do it! And then riffing off this suggestion we implemented From and made things a lot cleaner.

So I don't think it was a bad suggestion—the suggestion to consider it did indeed lead us to a nice solution!